SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp

```diff
@@ -0,0 +1,358 @@
+					&& DynamicReplicatedActor->AuthorityOnBeginPlay == 1
+					&& DynamicReplicatedActor->AuthorityOnTick == 1)
+				{
+					FinishStep();
+				}
+			}, 5.0f);
+
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Destroy"), FWorkerDefinition::Server(1), nullptr,
+			[this](ASpatialFunctionalTest* Test)
+			{
+				DynamicReplicatedActor->Destroy();
+				CrossServerSetDynamicReplicatedActor(nullptr);
+				FinishStep();
+			});
+	}
+
+	// Replicated Dynamic Actor Spawned On Different Server. Server 1 should have Authority on BeginPlay, Server 2 on Tick
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+				[this](ASpatialFunctionalTest* Test) {
+					ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
+						Server2Position, FRotator::ZeroRotator);
+					CrossServerSetDynamicReplicatedActor(Actor);
+					FinishStep();
+				});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Different Server - Verify Server 1 Has Authority on BeginPlay and Server 2 on Tick"), FWorkerDefinition::AllWorkers, nullptr,
+			nullptr,
+			[this](ASpatialFunctionalTest* Test, float DeltaTime) {
+				// Allow it to continue working in Native / Single worker setups.
+				int ExpectedAuthorityServer = GetNumberOfServerWorkers() > 1 ? 2 : 1;				
+				if (DynamicReplicatedActor != nullptr
+					&& DynamicReplicatedActor->AuthorityOnBeginPlay == 1
+					&& DynamicReplicatedActor->AuthorityOnTick == ExpectedAuthorityServer)
+				{
+					FinishStep();
+				}
+			},
+			5.0f);
+
+		// Now it's Server 2 destroying, since it has Authority over it.
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Destroy"), FWorkerDefinition::Server(2), nullptr,
+				[this](ASpatialFunctionalTest* Test)
+				{
+					DynamicReplicatedActor->Destroy();
+					CrossServerSetDynamicReplicatedActor(nullptr);
+					FinishStep();
+				});
+	}
+
+	// Non-replicated Dynamic Actor. Server 1 should have Authority.
+	{
+		AddStep(TEXT("Non-replicated Dynamic Actor - Spawn"), FWorkerDefinition::Server(1), nullptr,
+			[this](ASpatialFunctionalTest* Test)
+			{
+				// Spawning directly on Server 2, but since it's non-replicated it shouldn't migrate to Server 2.
+				DynamicNonReplicatedActor = GetWorld()->SpawnActor<ASpatialAuthorityTestActor>(Server2Position, FRotator::ZeroRotator);
+				FinishStep();
+			});
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Authority on Server 1"), FWorkerDefinition::Server(1), nullptr, nullptr,
+				[this](ASpatialFunctionalTest* Test, float DeltaTime)
+				{
+					if (DynamicNonReplicatedActor->AuthorityOnBeginPlay == 1
+						&& DynamicNonReplicatedActor->AuthorityOnTick == 1)
+					{
+						FinishStep();
+					}
+				});
+
+		float Timer = 0.5f;
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Dynamic Actor doesn't exist on others"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
```

Comment:
In this step, why do you use a Timer variable instead of relying on the test timeout to fail the step? Is it because you want to add the specific message if the test fails in the last lines of the step?

Reply:
Because that timer is an active wait, we wait that time and only then we do a check

Reply:
So basically you try to avoid situations like when the test could pass before the timer elapsed, even though it should not pass, right? I think that makes sense, thank you for the clarification 😄!

Reply:
Yes, this is the "worst kind of condition", because you need to actively wait to make sure that something never got propagated to you. We should avoid them in general, but some times we can't :-/ Not sure how we can better handle this, maybe at a certain point active wait tests can be separate, but it may require duplicating chunks of tests, so not great :-/

Reply:
I see, thank you for the info!